### PR TITLE
feat(windows): kill process gracefully

### PR DESCRIPTION
**Pull request by fahadmohammed01**
**Description of changes:**
Send ctrl-c event to process before forcefully killing the processes using taskkill.

**How was this change tested:**
mvn verify

**Any additional information or context required to review the change:**

**Checklist:**
 - [ ] Updated the README if applicable
 - [ ] Updated or added new unit tests
 - [ ] Updated or added new integration tests
 - [ ] Updated or added new end-to-end tests
 - [ ] If your code makes a remote network call, it was tested with a proxy
 - [ ] You confirm that the change is backwards compatible

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java [hunk 1]
```diff
@@ -140,6 +144,62 @@ public synchronized void close() throws IOException {
...
         }
 
+        try {
+            stopGracefully();
+        } finally {
+            stopForcefully();
+        }
+    }
+
+    private void stopGracefully() {
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:154]: break this out into a method which accepts a pid as input. In Tilo's change we'll know the pid

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java [hunk 2]
```diff
@@ -140,6 +144,62 @@ public synchronized void close() throws IOException {
...
+                if (!k32Ex.SetConsoleCtrlHandler(null, false)) {
+                    logger.error("SetConsoleCtrlHandler remove error {}", k32.GetLastError());
+                }
+            }
+        }
+
+        try {
+            if (sentConsoleCtrlEvent) {
+                process.waitFor(gracefulShutdownTimeout, TimeUnit.SECONDS);
+                logger.info("Process stopped gracefully: {}", pid);
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:196]: is this what we log for Unix? I don't want us to have extraneous logs

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java [hunk 3]
```diff
@@ -140,6 +144,62 @@ public synchronized void close() throws IOException {
...
+        } finally {
+            stopForcefully();
+        }
+    }
+
+    private void stopGracefully() {
+        int pid = Processes.newPidProcess(process).getPid();
+        // Global lock since we're messing with a shared resource (the console)
+        boolean sentConsoleCtrlEvent = false;
+        synchronized (Kernel32Ex.INSTANCE) {
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:158]: Please verify that after attaching and then freeing console, that _our_ application is still able to receive ctrl+c on windows. Just spawn a process during Nucleus launch and graceful kill it; then verify that ctrl+c still works. 
I am wondering if we need to re-attach at the end to our own PID's console or not.

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java [hunk 4]
```diff
@@ -140,6 +144,62 @@ public synchronized void close() throws IOException {
...
+            Kernel32 k32 = Kernel32.INSTANCE;
+            if (!k32.AttachConsole(pid)) {
+                logger.error("AttachConsole error {}", k32.GetLastError());
+                // Console already attached so we cannot signal it
+                if (k32.GetLastError() == ERROR_ACCESS_DENIED) {
+                    logger.info("AttachConsole failed for PID: {}, calling FreeConsole on it", pid);
+                    // Oddly, calling FreeConsole seems to make subsequent calls work?
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:165]: cleanup comment.

File: src/main/java/com/aws/greengrass/jna/Kernel32Ex.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.jna;
+
+import com.sun.jna.Library;
+import com.sun.jna.Native;
+import com.sun.jna.win32.W32APIOptions;
+
+public interface Kernel32Ex extends Library {
```
Comment (tilo-chen) [inline src/main/java/com/aws/greengrass/jna/Kernel32Ex.java:12]: This should be unused now?

File: src/main/java/com/aws/greengrass/jna/HandlerRoutine.java [hunk 6]
```diff
@@ -0,0 +1,12 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.jna;
+
+import com.sun.jna.win32.StdCallLibrary;
+
+public interface HandlerRoutine extends StdCallLibrary.StdCallCallback {
```
Comment (tilo-chen) [inline src/main/java/com/aws/greengrass/jna/HandlerRoutine.java:10]: unused?

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java [hunk 7]
```diff
@@ -118,6 +123,70 @@ public synchronized void close() throws IOException {
...
+                return;
+            }
+            Kernel32Ex k32Ex = Kernel32Ex.INSTANCE;
+            try {
+                if (!k32Ex.SetConsoleCtrlHandler(null, true)) {
+                    logger.error("SetConsoleCtrlHandler add error {}", k32.GetLastError());
+                    return;
+                }
+                if (k32.GenerateConsoleCtrlEvent(Wincon.CTRL_C_EVENT, 0)) {
+                    sentConsoleCtrlEvent = true;
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:166]: Try adding a slight sleep (1s to start) after generating the event. Other implementations have added a sleep "so that we don't kill ourselves", so maybe that is necessary.

File: src/main/java/com/aws/greengrass/jna/HandlerRoutine.java (hunk 6 above)
Comment (fahadmohammed01) [reply src/main/java/com/aws/greengrass/jna/HandlerRoutine.java:10]: It is

File: src/main/java/com/aws/greengrass/jna/Kernel32Ex.java (hunk 5 above)
Comment (fahadmohammed01) [reply src/main/java/com/aws/greengrass/jna/Kernel32Ex.java:12]: not any more

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java [hunk 8]
```diff
@@ -118,6 +124,76 @@ public synchronized void close() throws IOException {
...
+                } else {
+                    logger.error("GenerateConsoleCtrlEvent error {}", k32.GetLastError());
+                }
+            } finally {
+                if (!k32.FreeConsole()) {
+                    logger.error("FreeConsole error {}", k32.GetLastError());
+                }
+                // wait to ensure CtrlHandler is not enabled before the calling process gets the signal
+                try {
+                    Thread.sleep(2000);
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:178]: Does it need to be 2 seconds? Is that the only time you tried?

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java [hunk 9]
```diff
@@ -118,6 +124,76 @@ public synchronized void close() throws IOException {
...
+                }
+                if (!k32Ex.SetConsoleCtrlHandler(null, false)) {
+                    logger.error("SetConsoleCtrlHandler remove error {}", k32.GetLastError());
+                }
+            }
+        }
+
+        try {
+            if (sentConsoleCtrlEvent) {
+                process.waitFor(gracefulShutdownTimeout, TimeUnit.SECONDS);
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:189]: `gracefulShutdownTimeout-2` to account for the time we already waited

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java [hunk 10]
```diff
@@ -118,6 +124,76 @@ public synchronized void close() throws IOException {
...
+                logger.info("Process stopped gracefully: {}", pid);
+            }
+        } catch (InterruptedException ignore) {
+        }
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:193]: don't ignore it, throw it up or at least re-set the interrupted flag on the thread

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java [hunk 11]
```diff
@@ -118,6 +124,76 @@ public synchronized void close() throws IOException {
...
+                }
+            } finally {
+                if (!k32.FreeConsole()) {
+                    logger.error("FreeConsole error {}", k32.GetLastError());
+                }
+                // wait to ensure CtrlHandler is not enabled before the calling process gets the signal
+                try {
+                    Thread.sleep(2000);
+                } catch (InterruptedException ignored) {
+                }
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:180]: reset the interrupted flag

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java (hunk 8 above)
Comment (rbattle) [reply src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:178]: Do we need these timeouts (the 5s and 2s here) to be configurable?
If 2 seconds is not long enough, what happens?
If it causes a crash, we'll get customer tickets

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java (hunk 2 above)
Comment (rbattle) [reply src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:190]: this should be debug at highest

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java [hunk 12]
```diff
@@ -118,6 +124,76 @@ public synchronized void close() throws IOException {
...
+        }
+        int pid = process instanceof ProcessImplForWin32
+                ? ((ProcessImplForWin32) process).getPid() : Processes.newPidProcess(process).getPid();
+        boolean sentConsoleCtrlEvent = false;
+        synchronized (Kernel32.INSTANCE) {
+            Kernel32 k32 = Kernel32.INSTANCE;
+            // if the console is already attached, a second attach call will fail.
+            if (!k32.FreeConsole()) {
+                logger.error("FreeConsole error {}", k32.GetLastError());
+            }
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:148]: if the console is not attached, does this log an unnecesary error?

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java [hunk 13]
```diff
@@ -118,6 +124,76 @@ public synchronized void close() throws IOException {
...
+                        logger.error("FreeConsole error {}", k32.GetLastError());
+                    }
+                }
+                return;
+            }
+            Kernel32Ex k32Ex = Kernel32Ex.INSTANCE;
+            try {
+                if (!k32Ex.SetConsoleCtrlHandler(null, true)) {
+                    logger.error("SetConsoleCtrlHandler add error {}", k32.GetLastError());
+                    return;
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:165]: can we wrap this whole thing in a try/finally where we log a warning that the process was not stopped gracefully?

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java (hunk 8 above)
Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:178]: Not likely. The 5 second timeout is the same as on linux which should be configurable eventually, but I wouldn't do it right now.

I don't see how Windows would need more than 2 seconds to send a signal, but who knows.

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java (hunk 13 above)
Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:165]: There's a finally on the outside which does a force kill

File: src/main/java/com/aws/greengrass/util/Exec.java [hunk 14]
```diff
@@ -87,6 +87,7 @@
     private TimeUnit timeunit = TimeUnit.SECONDS;
     private Copier stderrc;
     private Copier stdoutc;
+    protected long gracefulShutdownTimeout = 5L;
```
Comment (rbattle) [inline src/main/java/com/aws/greengrass/util/Exec.java:90]: can this be a duration instead of a unit-less number?

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java (hunk 13 above)
Comment (rbattle) [reply src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:165]: thanks for verifying that. Resolving

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java (hunk 8 above)
Comment (rbattle) [reply src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:178]: ok - lets make sure we have a backlog to address this (configurating the timeout) then. 

I know @fahadmohammed01  tried 1 second and was still seeing the error occur so he switched to 2

Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:178]: wow...

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java (hunk 12 above)
Comment (fahadmohammed01) [reply src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:148]: FreeConsole only returns false when there is an error, calling  FreeConsole when the console is already detached will return true.

**Comment by tilo-chen** [comment]
There's a few `AttachConsole error 5` (ACCESS_DENIED) from IPCPubSubRemovalTest in windows workflow output. Really odd because we already have synchronization and we FreeConsole in finally.

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java [hunk 15]
```diff
@@ -118,6 +116,88 @@ public synchronized void close() throws IOException {
...
+                }
+                // Send Ctrl-C to all processes in the console
+                if (k32.GenerateConsoleCtrlEvent(Wincon.CTRL_C_EVENT, 0)) {
+                    sentConsoleCtrlEvent = true;
+                    // Wait to ensure CtrlHandler is not enabled before the calling process gets the signal
+                    try {
+                        Thread.sleep(2000);
+                    } catch (InterruptedException ignore) {
+                        logger.atWarn(STOP_GRACEFULLY_EVENT).log("CtrlHandler sleep interrupted");
+                        Thread.interrupted();  // clear the interrupted status
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:167]: This is not correct. `Thread.currentThread().interrupt()` is what we want; we need to re-interrupt the thread.
`interrupted()` returns you a boolean, it doesn't do anything

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java [hunk 16]
```diff
@@ -118,6 +116,88 @@ public synchronized void close() throws IOException {
...
+                holderProc.destroyForcibly();
+            }
+        }
+
+        if (sentConsoleCtrlEvent) {
+            try {
+                process.waitFor(gracefulShutdownTimeout, TimeUnit.SECONDS);
+                logger.debug("Process stopped gracefully: {}", pid);
+            } catch (InterruptedException ignore) {
+                Thread.interrupted();  // clear the interrupted status
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:195]: same

File: src/main/java/vendored/org/apache/dolphinscheduler/common/utils/process/ProcessImplForWin32.java [hunk 17]
```diff
@@ -752,13 +752,29 @@ private static void restoreIOEHandleState(WinNT.HANDLE[] stdIOE, Boolean[] inher
...
+                        // The Unicode version of this function, CreateProcessW, can modify the contents of this string
+                        // Therefore, this parameter cannot be a pointer to read-only memory
+                        // (such as a const variable or a literal string). If this parameter is a constant string,
+                        // the function may cause an access violation.
+                        WTypes.LPWSTR cmdWstr = new WTypes.LPWSTR(cmd);
+                        char[] cmdChars = cmdWstr.getPointer()
+                                .getCharArray(0, cmd.length() + 1);  // +1 terminating null char
+                        WTypes.LPWSTR lpEnvironment = envblock == null ? new WTypes.LPWSTR() : new WTypes.LPWSTR(envblock);
+                        createProcSuccess = Kernel32.INSTANCE.CreateProcessW(null, cmdChars, null, null,
+                                true, new WinDef.DWORD(PROCESS_CREATION_FLAGS), lpEnvironment.getPointer(), path, si, pi);
```
Comment (MikeDombo) [inline src/main/java/vendored/org/apache/dolphinscheduler/common/utils/process/ProcessImplForWin32.java:775]: where do we get the environment for running as the same user @tilo-chen ?

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java (hunk 15 above)
Comment (tilo-chen) [reply src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:167]: `interrupted` does clear the interrupted status according to https://docs.oracle.com/javase/8/docs/api/java/lang/Thread.html#interrupted--

File: src/main/java/vendored/org/apache/dolphinscheduler/common/utils/process/ProcessImplForWin32.java (hunk 17 above)
Comment (tilo-chen) [reply src/main/java/vendored/org/apache/dolphinscheduler/common/utils/process/ProcessImplForWin32.java:775]: We call ProcessBuilderForWin32.environment() in WindowsExec.createProcess. That initializes the environment, inheriting current process env vars.

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java (hunk 15 above)
Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:167]: We do not want to clear the interrupt flag though. Ignoring the interrupt is not right

Comment (tilo-chen) [reply src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:167]: Got it I misunderstood "reset"

Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:167]: Yes sorry, re-set as in set the flag again

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java (hunk 4 above)
Comment (fahadmohammed01) [reply src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java]: done

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java (hunk 2 above)
Comment (fahadmohammed01) [reply src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java]: changed to debug!

File: src/main/java/com/aws/greengrass/util/Exec.java (hunk 14 above)
Comment (fahadmohammed01) [reply src/main/java/com/aws/greengrass/util/Exec.java]: Updated to be duration

File: src/main/java/vendored/org/apache/dolphinscheduler/common/utils/process/ProcessImplForWin32.java [hunk 18]
```diff
@@ -983,6 +932,92 @@ private static void waitForTimeoutInterruptibly(WinNT.HANDLE handle, long timeou
      * Begin Amazon addition.
      */
 
+    /**
+     * Preparation for running process as another user. Populates the given extraInfo if applicable.
+     * @return environment block for CreateProcess* APIs
+     */
+    private String setupRunAsAnotherUser(String username, String password, Map<String, String> envMap,
```
Comment (MikeDombo) [inline src/main/java/vendored/org/apache/dolphinscheduler/common/utils/process/ProcessImplForWin32.java:939]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 942-957 into a separate method.

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java [hunk 19]
```diff
@@ -118,6 +116,89 @@ public synchronized void close() throws IOException {
...
 
+        try {
+            stopGracefully();
+        } finally {
+            stopForcefully();
+        }
+    }
+
+    @SuppressFBWarnings("SWL_SLEEP_WITH_LOCK_HELD")
+    private void stopGracefully() {
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:127]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 190-198 into a separate method.

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java [hunk 20]
```diff
@@ -118,6 +116,89 @@ public synchronized void close() throws IOException {
...
+        }
+
+        if (sentConsoleCtrlEvent) {
+            try {
+                long timeToWaitInSec = Math.max(gracefulShutdownTimeout.getSeconds() - 2, 0);
+                process.waitFor(timeToWaitInSec, TimeUnit.SECONDS);
+                logger.debug("Process stopped gracefully: {}", pid);
+            } catch (InterruptedException ignore) {
+                Thread.currentThread().interrupt();
+            }
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:197]: instead of catching this I think I would throw it up and just let the `finally` do the force kill

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java [hunk 21]
```diff
@@ -118,6 +116,89 @@ public synchronized void close() throws IOException {
...
+                    logger.atError(STOP_GRACEFULLY_EVENT)
+                            .log("Re-SetConsoleCtrlHandler error {}", k32.GetLastError());
+                }
+                holderProc.destroyForcibly();
+            }
+        }
+
+        if (sentConsoleCtrlEvent) {
+            try {
+                long timeToWaitInSec = Math.max(gracefulShutdownTimeout.getSeconds() - 2, 0);
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:192]: nice, thanks

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java (hunk 20 above)
Comment (fahadmohammed01) [reply src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:197]: We will have to catch InterruptedException in close() method anyway. Is there any difference between catching it here vs catching it in close() ?

Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:197]: I'm actually only worried about what I told you to add in re-interrupting the thread. That may mean that taskkill won't be started since I don't know that Java would start a process from an interrupted thread. But we really need to be sure that taskkill kills the process. We could also use `TerminateProcess` JNA call instead of taskkill since we own the process anyway.

Comment (tilo-chen) [reply src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:197]: We still rely on taskkill to terminate the whole process tree. Let's just throw it up?

Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:200]: Yep

File: src/integrationtests/resources/com/aws/greengrass/integrationtests/ipc/pubsub.yaml [hunk 22]
```diff
@@ -103,7 +103,7 @@ services:
     lifecycle:
       run:
         windows:
-          powershell -command sleep 1
+          powershell -command sleep 100
```
Comment (MikeDombo) [inline src/integrationtests/resources/com/aws/greengrass/integrationtests/ipc/pubsub.yaml:106]: why are these changes needed?

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java [hunk 23]
```diff
@@ -118,6 +116,96 @@ public synchronized void close() throws IOException {
...
+                // Must detach from current console before attaching to another
+                if (!k32.FreeConsole()) {
+                    logger.atError(STOP_GRACEFULLY_EVENT).log("FreeConsole error {}", k32.GetLastError());
+                }
+                // calling attachConsole right after a process is launched will fail with invalid handle error
+                // waiting a bit ensures that we can attach to a process that just got launched.
+                try {
+                    Thread.sleep(500);
+                } catch (InterruptedException ignored) {
+                }
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:157]: can this come outside the sync block?

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java [hunk 24]
```diff
@@ -118,6 +116,96 @@ public synchronized void close() throws IOException {
...
+                // Send Ctrl-C to all processes in the console
+                if (k32.GenerateConsoleCtrlEvent(Wincon.CTRL_C_EVENT, 0)) {
+                    sentConsoleCtrlEvent = true;
+                } else {
+                    logger.atError(STOP_GRACEFULLY_EVENT)
+                            .log("GenerateConsoleCtrlEvent error {}", k32.GetLastError());
+                }
+            } finally {
+                // Re-attach gg to original console and re-enable Ctrl-C
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:177]: These need to be guarded.

`finally` runs _after_ `return` so any of the early returns you have will still execute this finally which may not be desirable.

Comment (fahadmohammed01) [reply src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:177]: The first thing we do in the try block is freeconsole(), we need to reattach the console irrespective of where we return.

File: src/integrationtests/resources/com/aws/greengrass/integrationtests/ipc/pubsub.yaml (hunk 22 above)
Comment (fahadmohammed01) [reply src/integrationtests/resources/com/aws/greengrass/integrationtests/ipc/pubsub.yaml:106]: @tilo-chen

Comment (MikeDombo) [reply src/integrationtests/resources/com/aws/greengrass/integrationtests/ipc/pubsub.yaml:106]: If it was dead, why did we try to kill it? Sounds like something customers could run into too

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java (hunk 23 above)
Comment (fahadmohammed01) [reply src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java]: moved this wait to process creation logic

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java [hunk 25]
```diff
@@ -90,23 +95,23 @@ public Path which(String fn) {
...
         winPb.environment().putAll(environment);
-        winPb.user(username, new String(getPassword(username)));
-        return winPb.directory(dir).command(commands).start();
+        Process process = winPb.directory(dir).command(commands).start();
+        // calling attachConsole right after a process is launched will fail with invalid handle error
+        // waiting a bit ensures that we get the process handle from the pid
+        try {
+            Thread.sleep(500);
+        } catch (InterruptedException ignored) {
+        }
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:113]: interrupt the thread; very important in this location. `Thread.currentThread().interrupt()`

File: src/integrationtests/resources/com/aws/greengrass/integrationtests/ipc/pubsub.yaml (hunk 22 above)
Comment (tilo-chen) [reply src/integrationtests/resources/com/aws/greengrass/integrationtests/ipc/pubsub.yaml:106]: Right I should revert it. That actually covers more cases.
We do check `process.isAlive()` before the whole graceful kill. But the first AttachConsole can still fail with ERROR_ACCESS_DENIED.

File: src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java (hunk 25 above)
Comment (fahadmohammed01) [reply src/main/java/com/aws/greengrass/util/platforms/windows/WindowsExec.java:114]: Added

**Review by MikeDombo — APPROVED** [review]

**Review by rbattle — APPROVED** [review]
approving - but we need a way to determine if these timeouts are good enough or if they need to be configurable
